MBS-10916: showing AcousticBrainz info in the Recording sidebar

loujine opened the pull request:
# Problem
MBS-10916
Display AcousticBrainz info in Recording pages sidebar

# Solution
Use React hook to fetch submission count from AB API. If not zero, fetch low-level data (tonal and rhythm).
See screenshot in MBS-10916

# Checklist for author
Tested on local database with sample data, works as expected

# Action
Check that the number of requests/data transfer amount to the AcousticBrainz API is reasonable. Another solution I tested quickly would be to display the result in the figerprints tab instead: less API requests vs lower visibility of the information

loujine commented:
I didn't notice the problem, but my local instance is already very slow (several seconds to load a page), I will have to check.

loujine commented:
If I compare master/this branch on my laptop, a recording page loads in ~5 seconds in both cases, so I can't detect a difference.

The behavior should be the same as the FingerprintTable: I can see in firefox developer tools that the requests to AB are done after the rest (same time as the favicon) and takes <100ms.
If AB is down the fetch request fails immediately, and I don't think there's anything I want to catch() since I don't display anything if no data is found. If we want to have a "No data found in AB" displayed instead, I can prepare an error message. I don't think anything was caught for the FingerprintTable either

reosarevok commented on root/layout/components/sidebar/SidebarAcousticBrainz.js:
```diff
@@ -0,0 +1,81 @@
+ * later version: http://www.gnu.org/licenses/gpl-2.0.txt
+ */
+
+import React, {useEffect, useState} from 'react';
```
Please use `import * as React` (I'm changing the place you copied this one from in https://github.com/metabrainz/musicbrainz-server/pull/1580/files) for this one :)

mwiencek commented on root/layout/components/sidebar/SidebarAcousticBrainz.js:
```diff
@@ -0,0 +1,81 @@
+const SidebarAcousticBrainz = ({recording}: {recording: RecordingT}) => {
+  const [count, setCount] = React.useState(0);
+  const [isCountLoaded, setIsCountLoaded] = React.useState(false);
+  const [data, setData] = React.useState({});
+  const [isDataLoaded, setIsDataLoaded] = React.useState(false);
```
Might make sense to just make the default value for `data` `null` - then you can just check `data != null` to determine if it's loaded rather than relying on `isDataLoaded`.

loujine replied:
Good point, fixed

mwiencek commented on root/layout/components/sidebar/SidebarAcousticBrainz.js:
```diff
@@ -0,0 +1,81 @@
+      </h2>
+
+      <a className="external" href={`//acousticbrainz.org/${recording.gid}`}>
+        {l(`${count} submissions on AcousticBrainz`)}
```
To handle plural forms for numeric values we'd use `ln` instead of `l` here, and interpolate the count this way
```suggestion
        {texp.ln(
          '{count} submission on AcousticBrainz',
          '{count} submissions on AcousticBrainz',
          {count},
        )}
```

mwiencek commented on root/layout/components/sidebar/SidebarAcousticBrainz.js:
```diff
@@ -0,0 +1,81 @@
+/*
```
One more thing is that we'll need to move this file to somewhere under root/static/scripts/common/components/, perhaps root/static/scripts/common/components/sidebar/SidebarAcousticBrainz.js. This is mentioned briefly in HACKING.md, but basically scripts that run on the client (even if they're also used on the server) generally have to go under root/static/scripts/ so that translated strings work. The reason is that when we bundle translated strings as .js for the client, we want to only include strings actually used by client scripts to minimize the bundle size, as there are a lot of strings in server-side components that we don't need in the browser. So, we only extract client strings from root/static/scripts/.

loujine replied:
I moved the file to root/static/scripts/common/components/sidebar/
And I should definitely read again HACKING.md, that's on me
Thanks for the explanation

loujine commented:
@mwiencek : is there a way to install this branch on a test server and test the performance impact?

yvanzo commented:
@loujine: just by asking :genie: https://test.musicbrainz.org/recording/cdba2c78-2645-4286-a225-013f2735d415

loujine commented:
Rebased, fixed the tests, and the performance on test.mb.o looks good. So unless we want to change the threshold to display/hide the tonality key info, I think this is ready for production

alastair commented:
on my list for this week 👍

alastair commented on root/static/scripts/common/components/sidebar/AcousticBrainz.js:
```diff
@@ -0,0 +1,82 @@
+  }
+
+  function fetchAcousticBrainzData() {
+    const dataUrl = `//acousticbrainz.org/api/v1/${recording.gid}/low-level`;
```
you should use our low-data endpoint: https://acousticbrainz.readthedocs.io/api.html#get--api-v1-low-level
use `?features=tonal.key_key;tonal.key_scale`. I see that we don't support key_strength. I added it in https://github.com/metabrainz/acousticbrainz-server/pull/384

loujine replied:
Didn't know about that, thanks

alastair commented on root/static/scripts/common/components/sidebar/AcousticBrainz.js:
```diff
@@ -0,0 +1,82 @@
+
+          {data.tonal.key_strength > 0.5 ? (
+            <SidebarProperty className="acousticbrainz_key" label={l('Key:')}>
+              {`${data.tonal.key_key} ${data.tonal.key_scale}`}
```
just one thing to keep in mind here - we use the sharp variations of all keys, so Bb major will be reported as A# major. This means that some people might complain that it's "wrong". We could provide mappings here, so we always report "A#/Bb major" in these cases, for example.

I'm not sure what the best strength threshold is. I was reading [the paper that describes the algorithm](https://www.researchgate.net/publication/220669391_Tonal_Description_of_Polyphonic_Audio_for_Music_Content_Processing) but it doesn't show how the accuracy increases as the strength threshold is altered. I'll discuss this with my coworkers. For now 0.5 is probably a "safe" bet.

Do we want a way for people to complain? Or at the very least a popup that says "this is automatically generated and might be wrong"?

alastair commented:
I collected the strengths for about 150,000 of the items in the AB database: 
```
0.0: 0 (0.0%)
0.1: 48 (0.03%)
0.2: 535 (0.35%)
0.3: 2689 (1.74%)
0.4: 9618 (6.21%)
0.5: 22158 (14.3%)
0.6: 39251 (25.33%)
0.7: 46709 (30.15%)
0.8: 31339 (20.23%)
0.9: 2596 (1.68%)
```
e.g. 0.2 means all strengths from 0.2 to 0.3. Let's keep 0.5 as the threshold for now, but I wonder if 0.6 or 0.7 may be better to be more sure. The original paper gives an accuracy of only ~66% for getting the key right. I'm not sure exactly what the accuracy of the current version of the code is.

loujine commented:
Regarding key strength/confidence, I also thought about comparing all submissions and make sure they all found the same key. I had one example with strength > 0.5 and different keys for some submissions, I'll try to find it again

alastair commented:
Yeah, the different strengths (or key values) may be from users submitting incorrect recordings for a specific MBID. until we merge the similarity work (https://github.com/metabrainz/acousticbrainz-server/pull/353) we don't really have a way of deciding which of many different submissions is the best. This is coming.

dbogdanov commented:
Checking the distribution of key strength values, the `>0.5` is a reasonable start to get rid of the outliers. The estimation algorithm always finds the best match, even though the music may be atonal. 

So, it is really good to put it straight and mention in the UI that this is an automatic value to avoid any misunderstandings.

alastair commented on root/static/scripts/common/components/sidebar/AcousticBrainz.js:
```diff
@@ -32,16 +32,31 @@ const SidebarAcousticBrainz = ({recording}: {recording: RecordingT}) => {
+      return key;
+    }
+    return {
+      'A#': 'A♯/B♭',
```
👏

mwiencek commented on root/static/scripts/common/components/sidebar/AcousticBrainz.js:
```diff
@@ -0,0 +1,97 @@
+      'C#': 'C♯/D♭',
+      'D#': 'D♯/E♭',
+      'F#': 'F♯/G♭',
+      'G#': 'G♯/A♭',
```
I'd probably move this to a `const KEY_LABELS` or something after the imports, just to avoid the extra allocation on each call.

loujine replied:
Sure, I'll fix that

loujine commented:
Spent some time thinking about a better way to display the fact that the AB info is generated, and in the end I used <abbr> with the text "Automatic suggestion from entry #1/{count}". The <abbr> makes it obvious (I hope) that there's an explanation without adding too much text on the page itself. The key strength is also mentioned in the abbr title.

Also I stopped displaying the count (as "{count} submissions on AB") since I don't think the information is that relevant

yvanzo commented on root/static/scripts/common/components/sidebar/AcousticBrainz.js:
```diff
@@ -0,0 +1,107 @@
+  return count ? (
+    <>
+      <h2>
+        {l('Acoustic information')}
```
> Spent some time thinking about a better way to display the fact that the AB info is generated

@loujine, maybe just retitle the section with _analysis_:

```suggestion
        {l('Acoustic analysis')}
```

yvanzo commented:
Thank you @loujine for the changes, @alastair and @dbogdanov for the comments.

loujine commented:
Thanks everyone for the feedback!

yvanzo commented:
For the record, it has been amended through commit https://github.com/metabrainz/musicbrainz-server/commit/b18e3402cf2efc97ef52e5881bd6345913026516 to get string placeholders fit our localization system.
